skimage/transform/steerable.py

```diff
@@ -0,0 +1,406 @@
+
+        hi0dft = imdft * hi0mask
+        hi0 = np.fft.ifft2(np.fft.ifftshift(hi0dft))
+
+        coeff.insert(0, [hi0.real])
+
+        return coeff
+
+    def _build_pyr_level(self, lodft, log_rad, angle, Xrcos, Yrcos, ht):
+        """
+        Parameters
+        ----------
+        lodft : 2-D array
+            DFT matrix of the higher layer
+        log_rad, angle: 2-D array
+            helper to help create the DFT mask
+        Xrcos, Yrcos: 2-D array
+            represents the desired filter
+        ht: int
+            current level of the pyramid that are being built
+
+        Notes
+        -----
+        This method is called recursively until ht reaches 1.
+        """
+        if (ht <= 1):
+            lo0 = np.fft.ifft2(np.fft.ifftshift(lodft))
+            coeff = [[lo0.real]]
+
+        else:
+            Xrcos = Xrcos - 1
+
+            # ==================== Orientation bandpass =======================
+            himask = _point_op(log_rad, Yrcos, Xrcos)
+
+            lutsize = 1024
+            Xcosn = np.pi * \
+                np.array(range(-(2 * lutsize + 1), (lutsize + 2))) / lutsize
+            order = self.nbands - 1
+            const = np.power(2, 2 * order) * \
+                np.square(np.math.factorial(order)) / \
+                (self.nbands * np.math.factorial(2 * order))
+
+            alpha = (Xcosn + np.pi) % (2 * np.pi) - np.pi
+            Ycosn = 2 * np.sqrt(const) * \
+                np.power(np.cos(Xcosn), order) * (np.abs(alpha) < np.pi / 2)
+
+            orients = []
+
+            for b in range(self.nbands):
+                anglemask = _point_op(
+                    angle, Ycosn, Xcosn + np.pi * b / self.nbands)
+                banddft = np.power(np.complex(
+                    0, -1), self.nbands - 1) * lodft * anglemask * himask
+                band = np.fft.ifft2(np.fft.ifftshift(banddft))
+                orients.append(band)
+
+            # ================== Subsample lowpass ============================
+            dims = np.array(lodft.shape)
+
+            lostart = np.ceil((dims + 0.5) / 2) - \
+                np.ceil((np.ceil((dims - 0.5) / 2) + 0.5) / 2)
+            loend = lostart + np.ceil((dims - 0.5) / 2)
+
+            lostart = lostart.astype(np.int32)
+            loend = loend.astype(np.int32)
+
+            log_rad = log_rad[lostart[0]:loend[0], lostart[1]:loend[1]]
+            angle = angle[lostart[0]:loend[0], lostart[1]:loend[1]]
+            lodft = lodft[lostart[0]:loend[0], lostart[1]:loend[1]]
+
+            YIrcos = np.abs(np.sqrt(1 - Yrcos * Yrcos))
+            lomask = _point_op(log_rad, YIrcos, Xrcos)
+
+            lodft = lomask * lodft
+
+            coeff = self._build_pyr_level(
+                lodft, log_rad, angle, Xrcos, Yrcos, ht - 1)
+            coeff.insert(0, orients)
+
+        return coeff
+
+    def _recon_pyr_level(self, coeff, log_rad, Xrcos, Yrcos, angle):
```

Comment:
It looks like these methods share repeated code segments; is it possible to refactor out some common functionality?

Reply:
I agree that they're somewhat duplicate, but refactoring will cleanup a few lines only (since there's difference between build and recon here and there). Also the common code does not stand out as a separate function, i.e taking them out will make reading them more difficult since there are no more context.

Reply:
I agree with @stefanv. Despite the difference, there are many opportunities for simplification. For example, this code is identical between the two methods:

```python
           Xrcos = Xrcos - 1
           himask = _point_op(log_rad, Yrcos, Xrcos)
           lutsize = 1024
           Xcosn = np.pi * \
               np.array(range(-(2 * lutsize + 1), (lutsize + 2))) / lutsize
           order = self.n_bands - 1
           const = np.power(2, 2 * order) * \
               np.square(np.math.factorial(order)) / \
               (self.n_bands * np.math.factorial(2 * order=)
```

That is only the most obvious example. I think you should definitely look at improving the code sharing between the two. It will massively reduce the maintainability burden in the future.